backbone/src/main/java/org/researchstack/backbone/ui/step/layout/FingerprintStepLayout.java

```diff
@@ -0,0 +1,479 @@
+
+    public FingerprintStepLayout(Context context) {
+        super(context);
+    }
+
+    public FingerprintStepLayout(Context context, AttributeSet attrs) {
+        super(context, attrs);
+    }
+
+    public FingerprintStepLayout(Context context, AttributeSet attrs, int defStyleAttr) {
+        super(context, attrs, defStyleAttr);
+    }
+
+    @TargetApi(21)
+    public FingerprintStepLayout(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
+        super(context, attrs, defStyleAttr, defStyleRes);
+    }
+
+    @Override
+    public void initialize(Step step, StepResult result) {
+        validateAndSetStep(step);
+
+        InstructionStep instructionStep = this.step;
+
+        if (instructionStep.getTitle() == null) {
+            instructionStep.setTitle(getContext().getString(R.string.rsb_fingerprint_title));
+        }
+
+        if (instructionStep.getText() == null) {
+            instructionStep.setTitle(getContext().getString(R.string.rsb_fingerprint_text));
+        }
+
+        initFingerPrintManager();
+
+        super.initialize(step, result);
+
+        submitBar.setVisibility(View.GONE);
+
+        refreshDetailText(
+                getContext().getString(R.string.rsb_fingerprint_hint),
+                getContext().getColor(R.color.rsb_hint));
+
+        fingerprintAnimationRunnable = new Runnable() {
+            @Override
+            public void run() {
+                refreshDetailText(
+                        getContext().getString(R.string.rsb_fingerprint_hint),
+                        getContext().getResources().getColor(R.color.rsb_hint, null));
+                refreshImage(ResUtils.ANIMATED_FINGERPRINT, true);
+                postDelayed(fingerprintAnimationRunnable, DEFAULT_FINGERPRINT_ANIMATION_MILLIS);
+            }
+        };
+        post(fingerprintAnimationRunnable);
+    }
+
+    private void initFingerPrintManager() {
+        fingerprintManager = FingerprintManagerCompat.from(getContext());
+        // We already have a fingerprint set, and this steplayout will be dismissed once
+        // the callbacks are set
+        if (fingerprintStep.isCreationStep() && StorageAccess.getInstance().hasPinCode(getContext())) {
+            return;
+        }
+        if (initCipher()) {
+            cryptoObject = new FingerprintManagerCompat.CryptoObject(cipher);
+            startListening(cryptoObject);
+        }
+    }
+
+    @Override
+    protected void validateAndSetStep(Step step) {
+        super.validateAndSetStep(step);
+
+        if (!(step instanceof FingerprintStep)) {
+            throw new IllegalStateException("FingerprintStepLayout expects a FingerprintStep");
+        }
+
+        fingerprintStep = (FingerprintStep)step;
+    }
+
+    protected FingerprintManagerCompat.AuthenticationCallback authCallbacks = new FingerprintManagerCompat.AuthenticationCallback() {
+
+        static final int ERROR_MSG_TOO_MANY_ATTEMPTS = 7;
```

Comment:
Is this something that individual apps may want to set and should be passed in somehow?